File: web/typescript/gameassets/room/rect.ts
function lerp(a: number, b: number, alpha: number): number{
    return a + alpha * (b - a);
}

class Rect {
    position: Vector2;
    size: Vector2;

    constructor(abstractRect?: Rect){
        this.position = new Vector2(0, 0);
        this.size = new Vector2(0, 0);

        if(abstractRect){
            this.setPosition(abstractRect.position.x, abstractRect.position.y);
            this.setSize(abstractRect.size.x, abstractRect.size.y);
        }
    }

    setPosition(x: number, y: number){
        this.position.setPosition(x, y);
    }

    setSize(x: number, y: number){
        this.size.setPosition(x, y);
    }

    containsPoint(x: number, y: number){
        let pos = this.position;
        let size = this.size;

        if(
        x >= pos.x && x <= pos.x + size.x &&
        y >= pos.y && y <= pos.y + size.y
        ){
            return true;
        }
        return false;
    }

    getPosition(){
        return new Vector2(this.position.x, this.position.y);
    }

    getSize(){
        return new Vector2(this.size.x, this.size.y);
    }
}

File: internal/mathobjects/rect.go
package mathobjects

type Rect struct {
	Position Vector2[float64] `json:"position"`
	Size     Vector2[float64] `json:"size"`
}

func MakeRect(x, y, w, h float64) Rect {
	return Rect{
		Position: Vector2[float64]{x, y},
		Size:     Vector2[float64]{w, h},
	}
}

func (rect *Rect) Intersects(rect2 Rect) bool {
	if rect.Position.X+rect.Size.X <= rect2.Position.X ||
		rect.Position.X >= rect2.Position.X+rect2.Size.X ||
		rect.Position.Y+rect.Size.Y <= rect2.Position.Y ||
		rect.Position.Y >= rect2.Position.Y+rect2.Size.Y {
		return false
	}

	return true
}

func (rect *Rect) GetPosition() Vector2[float64] {
	return rect.Position
}

func (rect *Rect) GetSize() Vector2[float64] {
	return rect.Size
}
